sklearn/model_selection/_search_successive_halving.py

```diff
@@ -0,0 +1,895 @@
+                "no greater than max_resources."
+            )
+
+        if self.force_exhaust_resources and self.min_resources != 'auto':
+            raise ValueError(
+                'min_resources must be set to auto if force_exhaust_resources'
+                ' is True.'
+            )
+
+        self.min_resources_ = self.min_resources
+        if self.min_resources_ == 'auto':
+            if self.resource == 'n_samples':
+                cv = check_cv(self.cv, y,
+                              classifier=is_classifier(self.estimator))
+                n_splits = cv.get_n_splits(X, y, groups)
+
+                # please see https://gph.is/1KjihQe for a justification
+                magic_factor = 2
+                self.min_resources_ = n_splits * magic_factor
+                if is_classifier(self.estimator):
+                    n_classes = np.unique(y).shape[0]
+                    self.min_resources_ *= n_classes
+            else:
+                self.min_resources_ = 1
+
+        self.max_resources_ = self.max_resources
+        if self.max_resources_ == 'auto':
+            if not self.resource == 'n_samples':
+                raise ValueError(
+                    "max_resources can only be 'auto' if resource='n_samples'")
+            self.max_resources_ = _num_samples(X)
+
+        if self.min_resources_ > self.max_resources_:
+            raise ValueError(
+                'min_resources_={} is greater than max_resources_={}.'
+                .format(self.min_resources_, self.max_resources_)
+            )
+
+    def fit(self, X, y=None, groups=None, **fit_params):
+        """Run fit with all sets of parameters.
+
+        Parameters
+        ----------
+
+        X : array-like, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,) or (n_samples, n_output), optional
+            Target relative to X for classification or regression;
+            None for unsupervised learning.
+
+        groups : None
+            Groups are not supported
+
+        **fit_params : dict of string -> object
+            Parameters passed to the ``fit`` method of the estimator
+        """
+        self._check_input_parameters(
+            X=X,
+            y=y,
+            groups=groups,
+        )
+        super().fit(X, y=y, groups=None, **fit_params)
+        # Set best_score_: BaseSearchCV does not set it, as refit is a callable
+        self.best_score_ = (
+            self.cv_results_['mean_test_score'][self.best_index_])
+        return self
+
+    def _run_search(self, evaluate_candidates, X, y, **fit_params):
+        rng = check_random_state(self.random_state)
+
+        candidate_params = self._generate_candidate_params()
+        # Remove duplicates (may happen with random sampling)
+        candidate_params = set(tuple(d.items()) for d in candidate_params)
```

Comment:
The next line of code means the order of `candidate_params` is no longer deterministic. I'm not very comfortable with that, although I acknowledge that this is a randomised procedure, so determinism might not be the top of the wishlist.

If we can rely on the Python >= 3.6 dict ordering, then we can preserve order by replacing this with:
```suggestion
# We use a dict rather than a set for its deterministic iteration order
candidate_params = {tuple(d.items()): None for d in candidate_params}
```

Reply:
I don't remember why I did this. We don't seem to care about duplicate entries in RandomizedSearchCV so I'll assume we don't need to care about them in SH either. Tests seem to pass, though I might be missing something important?